### ISSUE TITLE
How can I integrate ContactView with ab_user?

### ISSUE THREAD

**Issue by allankliu** [dpgaspar/Flask-AppBuilder#681]
There are many examples designed around Contact/ContactGroup. And FAB has own ab_user/ab_role/ab_user_role tables. I am just confused about these seperate tables.

So, I have following questions:

1.  Can I merge Contact and ab_user tables ? and How ?
2. If not, why FAB perfer two user management system, and for what purposes ?
3. User registration, OAuth and DB authentication are working on FAB users, and how to apply to Contact/Group？
4. From schema of UserRegister, it is isolated from either systems. Why? and how can I integrate them together? Can I copy the information to contact table?

```
CREATE TABLE ab_register_user (
        id INTEGER NOT NULL,
        first_name VARCHAR(64) NOT NULL,
        last_name VARCHAR(64) NOT NULL,
        username VARCHAR(64) NOT NULL,
        password VARCHAR(256),
        email VARCHAR(64) NOT NULL,
        registration_date DATETIME,
        registration_hash VARCHAR(256),
        PRIMARY KEY (id),
        UNIQUE (username)
);
```

I am designing an IoT Web app. There are two major models: device and user. The user management should support DB authentication and OAuth methods.

**Comment by MichaelPereira** [comment]
the `ab_register_user` holds the information of users registered but not validated yet, you may or may not want to use it depending on your use case.

As far as extending `ab_user`, it's easy to do using sqlAlchemy. For example, this is how you can add a `slack_name` to the model by using your own model that extends `User`, 

```python

class CustomUser(User):
    __tablename__ = 'ab_user'
    slack_name = Column('slack_name', String(length=256))
    groups = relationship('Group', secondary=assoc_user_group, backref='users')


```

Then you replace the default SecurityManager with your own:

```python
from flask_appbuilder.security.sqla.manager import SecurityManager
from werkzeug.security import generate_password_hash

from models import CustomUser
from security_views import CustomUserDBModelView,\
    CustomRegisterUserDBView, CustomUserInfoEditView


class CustomSecurityManager(SecurityManager):
    user_model = CustomUser
    userdbmodelview = CustomUserDBModelView
    registeruserdbview = CustomRegisterUserDBView
    userinfoeditview = CustomUserInfoEditView

    # Overwrite parent's method to add slack_user name parameter
    # pylint: disable=too-many-arguments, arguments-differ, broad-except
    def add_user(self, username, first_name, last_name, email,
                 role, password='', hashed_password='', slack_name=None):
        """
            Generic function to create user
        """
        try:
            user = self.user_model()
            user.first_name = first_name
            user.last_name = last_name
            user.username = username
            user.email = email
            user.active = True
            user.roles.extend(role)
            user.slack_name = slack_name
            if hashed_password:
                user.password = hashed_password
            else:
                user.password = generate_password_hash(password)
            self.get_session.add(user)
            self.get_session.commit()
            logging.info(const.LOGMSG_INF_SEC_ADD_USER, username)
            return user
        except Exception as ex:
            logging.error(const.LOGMSG_ERR_SEC_ADD_USER, str(ex))
            return False
```

**Comment by allankliu** [comment]
Thanks, Michael, your reply gives a lot of detail. Before reading your reply, I have checked  examples/extendsecurity and added extra sec.py and MyUser class in models.py. 

I also added extra column in MyUser class, but after running it, the flask throws out an error, complaining table has no such column. It seems the project didn't clean up and update the schema properly. 

Therefore, so far MyUser class is just a subclass of User without any change. :) Anyway, the project is running anyway.

I will update my project after reading your code carefully.
